packages/token/src/erc20/extensions/erc4626/erc4626.cairo

```diff
@@ -0,0 +1,603 @@
+            self: @ComponentState<TContractState>, receiver: ContractAddress
+        ) -> Option::<u256> {
+            Option::None
+        }
+
+        fn withdraw_limit(
+            self: @ComponentState<TContractState>, owner: ContractAddress
+        ) -> Option::<u256> {
+            Option::None
+        }
+
+        fn redeem_limit(
+            self: @ComponentState<TContractState>, owner: ContractAddress
+        ) -> Option::<u256> {
+            Option::None
+        }
+    }
+
+    /// Allows contracts to hook logic into deposit and withdraw transactions.
+    /// This is where contracts can transfer fees.
+    pub trait ERC4626HooksTrait<TContractState> {
+        fn before_withdraw(ref self: ComponentState<TContractState>, assets: u256, shares: u256) {}
+        fn after_deposit(ref self: ComponentState<TContractState>, assets: u256, shares: u256) {}
+    }
+
+    //
+    // External
+    //
+
+    #[embeddable_as(ERC4626Impl)]
+    impl ERC4626<
+        TContractState,
+        +HasComponent<TContractState>,
+        impl Fee: FeeConfigTrait<TContractState>,
+        impl Limit: LimitConfigTrait<TContractState>,
+        impl Hooks: ERC4626HooksTrait<TContractState>,
+        impl Immutable: ImmutableConfig,
+        impl ERC20: ERC20Component::HasComponent<TContractState>,
+        +ERC20Component::ERC20HooksTrait<TContractState>,
+        +Drop<TContractState>
+    > of IERC4626<ComponentState<TContractState>> {
+        /// Returns the address of the underlying token used for the Vault for accounting,
+        /// depositing, and withdrawing.
+        fn asset(self: @ComponentState<TContractState>) -> ContractAddress {
+            self.ERC4626_asset.read()
+        }
+
+        /// Returns the total amount of the underlying asset that is “managed” by Vault.
+        fn total_assets(self: @ComponentState<TContractState>) -> u256 {
+            let this = starknet::get_contract_address();
+            IERC20Dispatcher { contract_address: self.ERC4626_asset.read() }.balance_of(this)
+        }
+
+        /// Returns the amount of shares that the Vault would exchange for the amount of assets
+        /// provided, in an ideal scenario where all the conditions are met.
+        fn convert_to_shares(self: @ComponentState<TContractState>, assets: u256) -> u256 {
+            self._convert_to_shares(assets, Rounding::Floor)
+        }
+
+        /// Returns the amount of assets that the Vault would exchange for the amount of shares
+        /// provided, in an ideal scenario where all the conditions are met.
+        fn convert_to_assets(self: @ComponentState<TContractState>, shares: u256) -> u256 {
+            self._convert_to_assets(shares, Rounding::Floor)
+        }
+
+        /// Returns the maximum amount of the underlying asset that can be deposited into the Vault
+        /// for the receiver, through a deposit call.
+        /// If the `LimitConfigTrait` is not defined for deposits, returns 2 ** 256 - 1.
```

Comment:
The implementation of LimitConfigTrait must always be available, even if it is the default. We should rephrase this. I don't make any specific suggestions since you will write a better sentence for sure.